### PR TITLE
[gui] [vulkan] A faster version of depth copy through ti.field/ti.ndarray (copy directly from vulkan to cuda/gpu/cpu)

### PR DESCRIPTION
**Pull request by Morcki**
## Usage
```python
window_shape = (720, 1080)
window = ti.ui.Window("Test for copy depth memory", window_shape)
canvas = window.get_canvas()
scene = ti.ui.Scene()
camera = ti.ui.make_camera()

# get the shape of window
w, h = window.get_window_shape()
# depth field/ndarray must have ti.f32 data type and 2d shape.
# and shape must be equal to the window shape
scene_depth = ti.ndarray(ti.f32, shape = (w, h))
# scene_depth = ti.field(ti.f32, shape = (w, h))

while window.running:
    render()
    canvas.scene(scene)
    window.get_depth_buffer_faster(scene_depth)
    window.show()
```
## Explanation
+ Q : ~~why use ndarray instead of field?~~ Now support both ndarray and field data structure.
+ A : ti.field struct has a complicated memory layout under unpacked mode(which is default mode for taichi), so the physical address of your index numbers may not directly to the right place(field memory must be padding to power of 2, and there may be some dirty data on the memory). As for ti.ndarray, it does not have this problem, because it's memory layout is what you see, and there is no other designs for it.

### CODE REVIEW

File: python/taichi/_kernels.py [hunk 1]
```diff
@@ -179,6 +182,19 @@ def ext_arr_to_matrix(arr: ndarray_type.ndarray(), mat: template(),
                     mat[I][p, q] = arr[I, p, q]
 
 
+@kernel
+def ext_arr_from_vulkan_layout_to_normal_layout(
```
Comment (YuCrazing) [inline python/taichi/_kernels.py:186]: Here we should add more comments to explain the purpose of this kernel's existence. And we should indicate that this kernel is only for internal usage.

File: python/taichi/ui/window.py [hunk 2]
```diff
@@ -126,6 +131,19 @@ def write_image(self, filename):
         """
         return self.window.write_image(filename)
 
+    def copy_depth_buffer_to_ndarray(self, depth: Ndarray):
+        """fetch the depth information of current scene to ti.Ndarray.
+           (support copy from vulkan to cuda/cpu which is a faster version)
+        Args:
+            depth (Ndarray): [window_width, window_height] carries depth information.
+        """
+        if not (isinstance(depth, Ndarray) and len(depth.shape) == 2):
```
Comment (YuCrazing) [inline python/taichi/ui/window.py:140]: I think we should check the shape of depth ndarrary here.

Comment (Morcki) [reply python/taichi/ui/window.py:140]: There is no function method to get resolution of window on python scope, so I judge this on the cpp part (window.cpp).

File: taichi/ui/common/window_base.h [hunk 3]
```diff
@@ -41,6 +42,8 @@ class WindowBase {
 
   virtual void write_image(const std::string &filename) = 0;
 
+  virtual void copy_depth_buffer_to_ndarray(taichi::lang::Ndarray *) = 0;
```
Comment (ailzhang) [inline taichi/ui/common/window_base.h:45]: I feel like it's possible to use `const taichi::lang::Ndarray& ndarray)` , could you give that a try?

File: python/taichi/_kernels.py [hunk 4]
```diff
@@ -179,6 +179,29 @@ def ext_arr_to_matrix(arr: ndarray_type.ndarray(), mat: template(),
...
+        vk_image: ndarray_type.ndarray(),
+        normal_image: ndarray_type.ndarray()):
+    '''extract ndarray of raw vulkan memory layout to normal memory
```
Comment (ailzhang) [inline python/taichi/_kernels.py:186]: Formatting for this docstring is a bit off - could you  please fix it?

File: python/taichi/_kernels.py [hunk 5]
```diff
@@ -179,6 +179,29 @@ def ext_arr_to_matrix(arr: ndarray_type.ndarray(), mat: template(),
...
+    layout. Explanation of why using ti.ndarray:
+        ti.field struct has a complicated memory layout under unpacked
```
Comment (ailzhang) [inline python/taichi/_kernels.py:188]: OOC: why do we want to mention `ti.field` in this docstring? I'd expect this one to explain what's vulkan memory layout   and normal layout means :P

File: taichi/ui/backends/vulkan/window.cpp [hunk 6]
```diff
@@ -102,6 +105,46 @@ void Window::write_image(const std::string &filename) {
...
+  // instead of accessing through the current SNode
+  if (depth_arr->ndarray_alloc_ == taichi::lang::kDeviceNullAllocation) {
+    TI_ERROR("Null Allocation for Depth-Ndarray!");
+  }
+
+  auto arr_dev_ptr = depth_arr->ndarray_alloc_.get_ptr();
+  renderer_->swap_chain().copy_depth_buffer_to_ndarray(arr_dev_ptr);
+
+  if (!config_.show_window) {
+    prepare_for_next_frame();
```
Comment (ailzhang) [inline taichi/ui/backends/vulkan/window.cpp:144]: OOC: if I call copy depth buffer twice without presenting, is it expected behavior to skip them?

File: python/taichi/ui/window.py [hunk 7]
```diff
@@ -126,6 +128,19 @@ def write_image(self, filename):
         """
         return self.window.write_image(filename)
 
+    def copy_depth_buffer_to_ndarray(self, depth: Ndarray):
```
Comment (ailzhang) [inline python/taichi/ui/window.py:131]:
```suggestion
    def get_depth_buffer(self, depth: Ndarray):
```

**Review by ailzhang — CHANGES_REQUESTED** [review]
Thanks for adding this feature! Added a few nits in comments. 
Would you mind adding some tests as well?

File: taichi/ui/backends/vulkan/window.cpp (hunk 6 above)
Comment (ailzhang) [reply taichi/ui/backends/vulkan/window.cpp:145]: sorry I was dumb :P

File: taichi/python/export_ggui.cpp [hunk 8]
```diff
@@ -287,10 +288,19 @@ struct PyWindow {
     window = std::make_unique<vulkan::Window>(prog, config);
   }
 
+  py::tuple get_window_shape() {
+    auto [w, h] = window->get_window_shape();
```
Comment (ailzhang) [inline taichi/python/export_ggui.cpp:292]: OOC: is there any difference between resolution saved in Appconfig and w&h saved in swap_chain? Looks like the former is easier to get?

**Review by ailzhang — COMMENTED** [review]
Thanks! I have a few small questions left inline, will leave the final review of this PR to @PENGUINLIONG ;)

File: taichi/ui/backends/vulkan/swap_chain.cpp [hunk 9]
```diff
@@ -45,6 +45,58 @@ void SwapChain::resize(uint32_t width, uint32_t height) {
...
+
+    BufferImageCopyParams copy_params;
+    copy_params.image_extent.x = w;
+    copy_params.image_extent.y = h;
+    copy_params.image_aspect_flag = VK_IMAGE_ASPECT_DEPTH_BIT;
+    cmd_list = stream->new_command_list();
+    cmd_list->image_to_buffer(depth_staging_buffer.get_ptr(), depth_allocation_,
+                              ImageLayout::transfer_src, copy_params);
+    cmd_list->image_transition(depth_allocation_, ImageLayout::transfer_src,
+                               ImageLayout::present_src);
```
Comment (PENGUINLIONG) [inline taichi/ui/backends/vulkan/swap_chain.cpp:78]: Depth image is never presented. Do you mean `ImageLayout::transfer_src`->`ImageLayout::depth_attachment`? Not a big issue on desktop platforms yet better to be corrected.

File: taichi/ui/backends/vulkan/swap_chain.cpp [hunk 10]
```diff
@@ -45,6 +45,58 @@ void SwapChain::resize(uint32_t width, uint32_t height) {
...
+
+  if (memcpy_cap == Device::MemcpyCapability::Direct) {
+    Device::AllocParams params{copy_size, /*host_wrtie*/ false,
+                               /*host_read*/ false, /*export_sharing*/ true,
+                               AllocUsage::Uniform};
+
+    auto depth_staging_buffer = device.allocate_memory(params);
+
+    device.image_transition(depth_allocation_, ImageLayout::present_src,
+                            ImageLayout::transfer_src);
```
Comment (PENGUINLIONG) [inline taichi/ui/backends/vulkan/swap_chain.cpp:68]: `ImageLayout::depth_attachment`?

**Review by PENGUINLIONG — CHANGES_REQUESTED** [review]
just noticed some improper image transitions but everything else lgtm, nice work

**Review by PENGUINLIONG — APPROVED** [review]
lgtm

**Pull request merged**